### ISSUE TITLE
[dtest-release]materialized_views_test::test_decommission_node_during_mv_insert_4_nodes - Exception: View mview.users_by_first_name not built

### ISSUE THREAD

**Issue by Annamikhlin** [scylladb/scylladb#20605]
failed on
https://jenkins.scylladb.com/job/scylla-master/job/dtest-release/681/testReport/junit/materialized_views_test/TestMaterializedViews/Run_Dtest_Parallel_Cloud_Machines___FullDtest___full_split005___test_decommission_node_during_mv_insert_4_nodes/

```
self = <materialized_views_test.TestMaterializedViews object at 0x7f9cd21dd040>

    def test_decommission_node_during_mv_insert_4_nodes(self):
        """Test removing node during MV inserts
        Test starts with a starting size 4 and removes one node during inserts into base table that cause to update materialized view as well
        (using cs_mv_profile.yaml profile).
        Validate the log has no errors.
        Issue #2783: there are mutation_write_timeout_exception in case starting size 4 and more
        """
>       self._run_node_failure_during_mv_stress_insert(rf=3, nodes=4, node_action="decommission", exclude_errors=["mutation_write_timeout_exception"])

materialized_views_test.py:231: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
materialized_views_test.py:307: in _run_node_failure_during_mv_stress_insert
    wait_for_view(cluster=self.cluster, session=session, ks="mview", view="users_by_first_name")
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

cluster = <ccmlib.scylla_cluster.ScyllaCluster object at 0x7f9ccf8ae1e0>
session = <cassandra.cluster.Session object at 0x7f9ccfc42780>, ks = 'mview'
view = 'users_by_first_name', raise_exception = True, timeout = 600, cl = 4

    def wait_for_view(cluster, session, ks, view, raise_exception=True, timeout=None, cl=None):  # noqa: PLR0913
        num_nodes = len(cluster.nodelist())
        num_alive = len([node for node in cluster.nodelist() if node.is_live()])
        if cl is None:
            cl = ConsistencyLevel.ONE if num_alive == 1 else ConsistencyLevel.QUORUM
        cl_name = ConsistencyLevel.value_to_name[cl]
        logger.debug(f"Waiting for view {ks}.{view} to finish building: num_nodes={num_nodes} num_alive={num_alive} cl={cl_name}")
    
        query = SimpleStatement(view_built_status_query(ks, view, "host_id,status"), consistency_level=cl)
    
        def _view_build_finished_on_live_nodes():
            status = defaultdict(set)
            entries = rows_to_list(session.execute(query))
            for entry in entries:
                status[entry[1]].add(entry[0])
            logger.debug(f"wait_for_view {ks}.{view}: status={status}")
            done = status["SUCCESS"]
            for node in cluster.nodelist():
                try:
                    if node.is_live() and not (UUID(node.hostid()) in done):
                        logger.debug(f"wait_for_view {ks}.{view}: node {node.hostid()}/{node.address()} is not done yet")
                        return False
                except NodetoolError:
                    # If we decomissioned a node with "nodetool decommission"
                    # the code above may temporarily think that node.is_alive()
                    # is still true, but node.hostid(), which calls nodetool,
                    # can fail with an exception. In this case we just need to
                    # consider this node non-live.
                    pass
            return True
    
        if not timeout:
            timeout = 600
            if hasattr(cluster, "scylla_mode") and cluster.scylla_mode == "debug":
                timeout = 900
    
        deadline = time.time() + timeout
        while time.time() < deadline:
            if _view_build_finished_on_live_nodes():
                return True
            time.sleep(5)
    
        error_msg = f"View {ks}.{view} not built"
        if raise_exception:
>           raise Exception(error_msg)
E           Exception: View mview.users_by_first_name not built

tools/tables_view_manager.py:716: Exception
```

full log: [dtest-gw4.log](https://github.com/user-attachments/files/17005844/dtest-gw4.log)

**Comment by mykaul** [comment]
Same in 6.2 - https://jenkins.scylladb.com/job/scylla-6.2/job/dtest-release-with-raft/1/testReport/materialized_views_test/TestMaterializedViews/Run_Dtest_Parallel_Cloud_Machines___FullDtest___full_split007___test_decommission_node_during_mv_insert_4_nodes/

**Comment by kostja** [comment]
@piodul most likely the test needs to be updated to never lose raft quorum, could you please assign someone to take a look?

**Comment by piodul** [comment]
@kostja It's not a problem with quorum being lost, this test starts with 4 nodes and shuts down only one during its course.

@Jadw1 please take a look - let's start by reproducing this test locally.

**Comment by Jadw1** [comment]
I was able to reproduce the issue locally, simply by running the dtest. 
I've triggered the error on release and dev modes, I didn't checked debug mode yet.

Unfortunately, the test takes about 15 minutes to complete, so it isn't convenient to debug.

No conclusion what can be a cause yet, since I don't know materialized views.

**Comment by Annamikhlin** [comment]
seen again on https://jenkins.scylladb.com/job/scylla-master/job/dtest-release/685/testReport/junit/materialized_views_test/TestMaterializedViews/Run_Dtest_Parallel_Cloud_Machines___FullDtest___full_split005___test_decommission_node_during_mv_insert_4_nodes/

**Comment by Jadw1** [comment]
Content of view build status tables:

system_distributed.view_build_status
```
keyspace_name    view_name            host_id                               status
---------------  -------------------  ------------------------------------  --------
mview            users_by_first_name  1c8c6bc9-51e9-4fa0-aec4-cb4bac6c266b  SUCCESS
mview            users_by_first_name  28e69781-c50c-4f8b-aad5-802664771571  SUCCESS
mview            users_by_first_name  8190b655-2eb4-40f7-a3bb-84f61915ff1f  SUCCESS
mview            users_by_first_name  b2ecfa0d-5050-4eb3-8243-d504f85084ab  SUCCESS
mview            users_by_last_name   1c8c6bc9-51e9-4fa0-aec4-cb4bac6c266b  SUCCESS
mview            users_by_last_name   28e69781-c50c-4f8b-aad5-802664771571  SUCCESS
mview            users_by_last_name   8190b655-2eb4-40f7-a3bb-84f61915ff1f  SUCCESS
mview            users_by_last_name   b2ecfa0d-5050-4eb3-8243-d504f85084ab  SUCCESS
```

system.view_build_status_v2
```
keyspace_name    view_name            host_id                               status
---------------  -------------------  ------------------------------------  --------
mview            users_by_first_name  1c8c6bc9-51e9-4fa0-aec4-cb4bac6c266b  SUCCESS
mview            users_by_first_name  28e69781-c50c-4f8b-aad5-802664771571  SUCCESS
mview            users_by_first_name  8190b655-2eb4-40f7-a3bb-84f61915ff1f  SUCCESS
mview            users_by_first_name  b2ecfa0d-5050-4eb3-8243-d504f85084ab  SUCCESS
mview            users_by_last_name   1c8c6bc9-51e9-4fa0-aec4-cb4bac6c266b  SUCCESS
mview            users_by_last_name   28e69781-c50c-4f8b-aad5-802664771571  SUCCESS
mview            users_by_last_name   8190b655-2eb4-40f7-a3bb-84f61915ff1f  SUCCESS
mview            users_by_last_name   b2ecfa0d-5050-4eb3-8243-d504f85084ab  SUCCESS
```

**Comment by Jadw1** [comment]
The problem is that one host_id in `system_distributed.view_build_status` table is wrong.

For instance:
Host ids of all nodes at the beginning of the test:
```
48c9685e-9f5e-4270-8e89-0d8d2598e347
57467175-6e95-4f3f-be05-3ed3710ef3dd
76daa6e7-5ddf-4529-aac7-b57c24a57fc3
bff2c11d-cbbb-4ef9-833b-a7df2fce0ca4
```

This function is used to check if the view was build on all live nodes: https://github.com/scylladb/scylla-dtest/blob/next/tools/tables_view_manager.py#L682-L701
It does a query to `system_distributed.view_build_status` table, which content is:
```
keyspace_name    view_name            host_id                               status
---------------  -------------------  ------------------------------------  --------
mview            users_by_first_name  411799e3-7a1f-47ff-9a91-dbdc1d74f71d  SUCCESS
mview            users_by_first_name  48c9685e-9f5e-4270-8e89-0d8d2598e347  SUCCESS
mview            users_by_first_name  76daa6e7-5ddf-4529-aac7-b57c24a57fc3  SUCCESS
mview            users_by_first_name  bff2c11d-cbbb-4ef9-833b-a7df2fce0ca4  SUCCESS
mview            users_by_last_name   411799e3-7a1f-47ff-9a91-dbdc1d74f71d  SUCCESS
mview            users_by_last_name   48c9685e-9f5e-4270-8e89-0d8d2598e347  SUCCESS
mview            users_by_last_name   76daa6e7-5ddf-4529-aac7-b57c24a57fc3  SUCCESS
mview            users_by_last_name   bff2c11d-cbbb-4ef9-833b-a7df2fce0ca4  SUCCESS
```

As you can see, for each view, there are 4 success entries, so it looks like the view was build correctly, but one of the host id doesn't match any node's host id.

In this example:
- there is a node `57467175-6e95-4f3f-be05-3ed3710ef3dd` without corresponding view status 
- there is a view build status on host id `411799e3-7a1f-47ff-9a91-dbdc1d74f71d` without corresponding node

**Comment by piodul** [comment]
Because the test fails reliably, I decided to run git bisect overnight, and it led me to this [commit](https://github.com/scylladb/scylladb/commit/446ad3c184db16f8f13cc584a24c36c6d2b84827):

```
446ad3c184db16f8f13cc584a24c36c6d2b84827 is the first bad commit
commit 446ad3c184db16f8f13cc584a24c36c6d2b84827
Author: Michael Litvak <michael.litvak@scylladb.com>
Date:   Thu Aug 8 15:51:51 2024 +0300

    view: delete node rows from view_build_status on node removal
    
    When a node is removed we want to clean its rows from the
    view_build_status table.
    Now when removing a node and generating the topology state update, we
    generate also the mutations to delete all the possible rows belonging to
    the node from the table.
```

**Comment by piodul** [comment]
TLDR: the test is broken, but it used to work because Scylla didn't use to clean up entries in view_build_status for nodes that left the cluster (which we now do, and is the correct thing to do).

Unlike what I wrote before, the test does not really shut down any node, it decommissions one of the initial four nodes and then adds one more node. Then, it waits until views are built on all "live" nodes.

This is the relevant part of the `wait_for_view` function:

```python
    def _view_build_finished_on_live_nodes():
        status = defaultdict(set)
        entries = rows_to_list(session.execute(query))
        for entry in entries:
            status[entry[1]].add(entry[0])
        logger.debug(f"wait_for_view {ks}.{view}: status={status}")
        done = status["SUCCESS"]
        for node in cluster.nodelist():
            try:
                if node.is_live() and not (UUID(node.hostid()) in done):
                    logger.debug(f"wait_for_view {ks}.{view}: node {node.hostid()}/{node.address()} is not done yet")
                    return False
            except NodetoolError:
                # If we decomissioned a node with "nodetool decommission"
                # the code above may temporarily think that node.is_alive()
                # is still true, but node.hostid(), which calls nodetool,
                # can fail with an exception. In this case we just need to
                # consider this node non-live.
                pass
        return True

    if not timeout:
        timeout = 600
        if hasattr(cluster, "scylla_mode") and cluster.scylla_mode == "debug":
            timeout = 900

    deadline = time.time() + timeout
    while time.time() < deadline:
        if _view_build_finished_on_live_nodes():
            return True
        time.sleep(5)
```

In a loop, the function goes over nodes from the `cluster.nodelist()` and checks their view building status in the contents of the `view_build_status` table.

First and foremost, the decommissioned node is included in the `cluster.nodelist()` - not sure if this is intended, but considering that the node is only decommissioned and not shut down it does make some sense.

Second, it looks like `node.is_live()` check was introduced to filter out the decommissioned node, but it does not work. The `node` is a `ccmlib` object; I looked at `is_live()`'s implementation:

```python
    def is_live(self):
        """
        Return true if the node is live (it's run and is not decommissioned).
        """
        self.__update_status()
        return self.status == Status.UP
```

```python
    def __update_status(self):
        if self.pid is None:
            if self.status == Status.UP or self.status == Status.DECOMMISSIONED:
                self.status = Status.DOWN
            return

        old_status = self.status

        # os.kill on windows doesn't allow us to ping a process
        if common.is_win():
            self.__update_status_win()
        else:
            try:
                os.kill(self.pid, 0)
            except OSError as err:
                if err.errno == errno.ESRCH:
                    # not running
                    if self.status == Status.UP or self.status == Status.DECOMMISSIONED:
                        self.status = Status.DOWN
                elif err.errno == errno.EPERM:
                    # no permission to signal this process
                    if self.status == Status.UP or self.status == Status.DECOMMISSIONED:
                        self.status = Status.DOWN
                else:
                    # some other error
                    raise err
            else:
                if self.status == Status.DOWN or self.status == Status.UNINITIALIZED:
                    self.status = Status.UP

        if not old_status == self.status:
            if old_status == Status.UP and self.status == Status.DOWN:
                self.pid = None
            self._update_config()
```

It pings the process, and if it responds then the node is marked as `UP` if it was `DOWN` or `UNINITIALIZED` before. It looks like it should be marked as `DECOMMISSIONED` by this point - the only place where it happens is in `Node.decommission()`:

```python
    def decommission(self):
        self.nodetool("decommission")
        self.status = Status.DECOMMISSIONED
        self._update_config()
```

However, the test does not call it, but rather calls `node.nodetool("decommission")` directly, bypassing the assignment of the `DECOMMISSIONED` status:

```python
    def _node_action_with_delay(self, action, node, delay=0, wait=True, wait_other_notice=True, other_nodes=None, gently=True):  # noqa: PLR0913
        """
        :param action: expected values: stop, remove
        :param action: str
        """
        if action not in ["stop", "remove", "decommission", "restart"]:
            assert False, "Unsupported node action"

        if delay:
            logger.debug(f"Sleep for {delay} seconds")
            time.sleep(delay)

        logger.debug(f"START: {action} node {node.name}")
        if action == "stop":
            node.stop(wait=wait, wait_other_notice=wait_other_notice, other_nodes=other_nodes, gently=gently)
        elif action == "restart":
            node.stop(wait=wait, wait_other_notice=wait_other_notice, other_nodes=other_nodes, gently=gently)
            time.sleep(delay if delay else 1)
            node.start(wait_other_notice=wait_other_notice)
        elif action == "remove":
            remove_node(self.cluster, node, wait_other_notice=wait_other_notice, other_nodes=other_nodes)
        else:
            new_node_index = len(self.cluster.nodelist()) + 1
            node.nodetool(action)
            if action == "decommission":
                logger.debug("START add new node")
                self._add_new_node(new_node_index=new_node_index)
                logger.debug("FINISH add new node")

        logger.debug(f"FINISH: {action} node {node.name}")
```

I think this is a bug/footgun in ccmlib that `node.nodetool("decommission")` does not mark the node as decommissioned afterwards, but I'm afraid that fixing this would have effect on other tests, so I'll start with a less intrusive fix and I'll patch the `_node_action_with_delay` test function to call `node.decommission()` explicitly. After release, we can consider patching `node.nodetool()` function.

**Comment by piodul** [comment]
Sent a fix to dtests: https://github.com/scylladb/scylla-dtest/pull/4977
